core/src/main/java/org/apache/calcite/rel/rules/FilterProjectTransposeRule.java

```diff
@@ -155,8 +158,12 @@ public void onMatch(RelOptRuleCall call) {
     final RelBuilder relBuilder = call.builder();
     RelNode newFilterRel;
     if (copyFilter) {
-      newFilterRel = filter.copy(filter.getTraitSet(), project.getInput(),
-          simplifyFilterCondition(newCondition, call));
+      final RelNode input = project.getInput();
+      final RelTraitSet traitSet = filter.getTraitSet()
+          .replaceIfs(RelCollationTraitDef.INSTANCE,
```

Comment:
What about the RelDistribution trait? Will it have the same problem?

Reply:
Fixed.
Well spotted, I forgot to add it.

Reply:
wouldn't it be better to use the relbuilder to create a new filter instead since it would defer to *Filter#create (which already has the correct logic), or fix Filter#copy method instead?

Reply:
@laurentgo, this is already possible, by using `copyFilter=false` when creating the rule.

Reply:
The way the rule currently is, If it matches an EnumerableFilter and EnumerableProject with `copy==false` it will create logical equivalents and not enumerable operators (provided that the RelBuilderFactory is the logical one).

Reply:
This rule can match both logical and physical operator. But why does it match physical operator? It doesn't make sense. We already transposed filter & project on logical operator, it would be redundant to apply this rule on physical operator again.

Reply:
The mechanistic answer to "why" is because its operands are defined on Project.class and Filter.class, and therefore matches any sub-type of these.

You can if you wish create other instances of the rule that match, say, LogicalProject.class and LogicalFilter.class.

Reply:
I am not making myself clear. I mean do we need to apply this rule on physical operator since it has already been applied on logical operators?

Reply:
Probably you don't.

But perhaps you have some physical optimizations that produce new physical filters that need to be pushed down.

Or perhaps you have pulled up filters in order to recognize materialized views.

Only you can say, based on the phases of your optimization process and the rules enabled in each. Safest thing is to disable the rule, and enable it if you start seeing sub-optimal plans.

Reply:
AggregateStarTableRule, FilterAggregateTransposeRule, AggregateProjectMergeRule, FilterJoinRule, FilterMergeRule, ProjectMergeRule... Their default instances all matches both logical and physical operators, but they all use RelBuilder to build logical operators. Some rules like AggregateJoinTransposeRule, the default instance only matches logical operators. There is no consistency among those rules. IMHO, rather than fix it, we should just remove or deprecate the option of copyFilter, let the rule always match logical operators.

Reply:
I don't like the inconsistency either but I am not sure that the rules you mentioned really do what they are supposed to. Matching an EnumerableFilter and creating a LogicalFilter does not seem like the expected behavior.  This behavior may lead to problems like the one mentioned in CALCITE-2223. @vlsi started a discussion in the dev list (subject "Principle of least astonishment when dealing with Rules and RelBuidlerFactory" but unfortunately we didn't conclude to how we should deal with such rules. I think we could remove the `copyFilter`, `copyProject` but then the default rule (`INSTANCE`) should match only `LogicalFilter.class` and `LogicalProject.class`.

Reply:
I don't like the inconsistency either.

It would be nice if the default instance of these rules matched Filter, Project and Aggregate of any convention and automatically transformed to Filter, Project and Aggregate of the same convention. I know that's asking a lot. But it would be a user-friendly default for people who don't want to go to the effort of building rule sets and multi-phase planning.

People who are building complex planners would, I'm sure, create instances of those rules that match only one convention.

I'd also love to deprecate the `RelNode.copy` method and all of its variants in sub-classes. That is the reason why these rules behave inconsistently between single-convention and multi-convention mode.

Could we add a method `RelBuilder RelBuilder.withConvention(Convention)`? Thus rather than

```
filter.copy(traitSet, input, newCondition)
```

you'd write

```
relBuilder.withConvention(filter.getConvention)
  .push(input)
  .filter(newCondition)
  .build()
```

Reply:
But with the default INSTANCE, the RelBuilder comes from logical (or some other convention) RelBuilder factory, so using a logical builder to build a physical (another convention) operator still doesn't seem right. I don't think we can cater for everyone with a single default rule INSTANCE, if the default one doesn't satisfy his/her needs, he/she should create a new one to match operator with expected convention and relbuilder.

Even for those who don't want to sweat on building rule sets or multi-phase planning, I still don't see why the rule should match both logical and physical operators. Say we have
1.
```
LogicalFilter
  +-- LogicalProject
```
, with transpose rule, we will have
2.
```
LogicalProject
  +-- LogicalFilter
```
, with implementation rules, we will have
3.
```
EnumerableFilter
  +-- EnumerableProject
```
and
4.
```
EnumerableProject
  +-- EnumerableFilter
```

Again, with transpose rules (it can match physical operators), it matches expression 3, creates new relational expression like 4, and matches 4, creates relational expression same with 3. It is not only redundant, but will also increase the optimization time. 

This is also why in our system, almost all the default Calcite rule INSTANCEs can't be used, we have to create new ones.

Reply:
As a power-user, you're definitely not in the target audience for the default instances. It's not at all surprising that you're creating new instances.

It may not be clear, but the line of code `relBuilder.withConvention(filter.getConvention())` is creating a new `RelBuilder` with a factory that creates `LogicalFilter`, `LogicalProject`, etc. if `filter` is logical, creates `EnumerableFilter`, `EnumerableProject`, etc. if `filter` is enumerable, creates `JdbcFilter`, `JdbcProject`, etc. if `filter` is JDBC convention, and so forth. So, it doesn't matter that the `RelBuilder` starts off logical, because the rule switches to a new one according to the current convention.

Reply:
I agree that flexibility is nice but with the rules I would be more conservative. I would prefer that our default instances all (excluding converters and other special rules) match and transform from/to Logical operators. I think that people who bother to build their own ruleset are capable of instantiating the rules properly if needed. Those who don't are more or less those using the Volcano planner as a black box with the default ruleset. My opinion may be a bit biased since I am also in the same position as @hsyuan.